### DISCUSSION TITLE
Async support for custom module's preRegisterCheck method

### DISCUSSION THREAD

**Discussion by t7tran** [directus/directus#6103] [Feature Requests]
What do you guys think of adding `async` support to `preRegisterCheck()` method in custom modules?

In our cases, we need to prefetch some resources depending on user's preferences. @rijkvanzanten this includes our workaround for #3075.

The change is simply in `modules/register.ts` as follow:

```
if (mod.preRegisterCheck) {
	if (mod.preRegisterCheck.constructor.name === "AsyncFunction")
		return await mod.preRegisterCheck(userStore.state.currentUser, permissionsStore.state.permissions);
	return mod.preRegisterCheck(userStore.state.currentUser, permissionsStore.state.permissions);
}
```

I'm happy to push a PR for this if the proposal looks OK to you guys.

**Comment by rijkvanzanten** [comment]
Sounds good to me! We'll have to make sure the app defers it's actual mounting until all the checks have completed though, otherwise the user can end up getting a 404 for valid route on refresh / login

**Reply by t7tran** [reply]
That's exactly the kind of insights I need. I'm just unsure I understand what you mean. If the register function is modified as suggested, it would block the thread until the execution completes, as if an infinitive loop occurs in the synchronous version of `preRegisterCheck` (which I verified).

**Reply by rijkvanzanten** [reply]
Yup! You're correct. I already built it in a way where just plugging in the `await` in from of preRegisterCheck does the trick 🙂 I wasn't sure if that was in an async function currently or not 👍🏻 

You don't have to explicitly check for `AsyncFunction`, as `await` on a regular function will resolve immediately 👍🏻

**Reply by t7tran** [reply]
Thanks for the tip. PR #6118 is submitted. I had to convert the filter into for loop to deal with `async` invocations.
